cloudfoundry-client-lib/src/main/java/org/cloudfoundry/client/lib/adapters/RawApplicationLog.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2013-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.cloudfoundry.client.lib.adapters;
+
+import org.cloudfoundry.client.lib.domain.ApplicationLog;
+import org.cloudfoundry.client.lib.domain.Derivable;
+import org.cloudfoundry.client.lib.domain.ImmutableApplicationLog;
+import org.cloudfoundry.doppler.LogMessage;
+import org.cloudfoundry.doppler.MessageType;
+import org.immutables.value.Value;
+
+import java.util.Date;
+import java.util.concurrent.TimeUnit;
+
+@Value.Immutable
+public abstract class RawApplicationLog implements Derivable<ApplicationLog> {
+
+    @Value.Parameter
+    public abstract LogMessage getLogMessage();
+
+    @Override
+    public ApplicationLog derive() {
+        LogMessage logMessage = getLogMessage();
+        return ImmutableApplicationLog.builder()
+                                      .applicationGuid(logMessage.getApplicationId())
+                                      .message(logMessage.getApplicationId())
+                                      .timestamp(fromLogTimestamp(logMessage))
+                                      .messageType(fromLogMessageType(logMessage))
+                                      .sourceId(logMessage.getSourceInstance())
+                                      .sourceName(logMessage.getSourceType())
+                                      .build();
+    }
+
+    private static Date fromLogTimestamp(LogMessage logMessage) {
+        return new Date(TimeUnit.NANOSECONDS.toMillis(logMessage.getTimestamp()));
+    }
+
+    private static ApplicationLog.MessageType fromLogMessageType(LogMessage logMessage) {
+        return logMessage.getMessageType() == MessageType.OUT ? ApplicationLog.MessageType.STDOUT :
```

Comment:
Use `org.cloudfoundry.client.lib.adapters.RawCloudEntity.parseEnum(Enum<?>, Class<E>)` instead of this method.

Reply:
Nvm, it wouldn't work for these two enums.